File: _CÓDIGO FONTE/Capítulo 3/02/lista_encadeada.js
class Nó {
    constructor(dado) {
        this.dado = dado;
        this.proximo = null;
    }
}

class ListaEncadeada {
    constructor() {
        this.cabeca = null;
    }

    adicionar(dado) {
        const novoNo = new Nó(dado);

        if (!this.cabeca) {
            this.cabeca = novoNo;
            return;
        }

        let atual = this.cabeca;
        while (atual.proximo) {
            atual = atual.proximo;
        }

        atual.proximo = novoNo;
    }

    inserirNoInicio(dado) {
        const novoNo = new Nó(dado);
        novoNo.proximo = this.cabeca;
        this.cabeca = novoNo;
    }

    inserirNaPosicao(posicao, dado) {
        const novoNo = new Nó(dado);
        let atual = this.cabeca;
        let anterior = null;
        let indice = 0;

        if (posicao === 0) {
            this.inserirNoInicio(dado);
            return;
        }

        while (atual && indice < posicao) {
            anterior = atual;
            atual = atual.proximo;
            indice++;
        }

        novoNo.proximo = atual;
        anterior.proximo = novoNo;
    }

    remover(dado) {
        if (!this.cabeca) {
            return;
        }

        if (this.cabeca.dado === dado) {
            this.cabeca = this.cabeca.proximo;
            return;
        }

        let atual = this.cabeca;
        let anterior = null;

        while (atual && atual.dado !== dado) {
            anterior = atual;
            atual = atual.proximo;
        }

        if (!atual) {
            return;
        }

        anterior.proximo = atual.proximo;
    }
}

// Criação da lista encadeada e testes dos métodos
var lista = new ListaEncadeada();
lista.adicionar(1);
lista.adicionar(10);
lista.adicionar(3);
lista.inserirNoInicio(2);
lista.remover(10);
lista.inserirNaPosicao(1, 3);


console.log(lista);

File: _CÓDIGO FONTE/Capítulo 3/02/lista_encadeada.py
class Nó:
    def __init__(self, dado):
        self.dado = dado
        self.proximo = None

class ListaEncadeada:
    def __init__(self):
        self.cabeca = None

    def adicionar(self, dado):
        novo_no = Nó(dado)

        if not self.cabeca:
            self.cabeca = novo_no
            return

        atual = self.cabeca
        while atual.proximo:
            atual = atual.proximo

        atual.proximo = novo_no

    def inserir_no_inicio(self, dado):
        novo_no = Nó(dado)
        novo_no.proximo = self.cabeca
        self.cabeca = novo_no

    def inserir_na_posicao(self, posicao, dado):
        novo_no = Nó(dado)
        atual = self.cabeca
        anterior = None
        indice = 0

        if posicao == 0:
            self.inserir_no_inicio(dado)
            return

        while atual and indice < posicao:
            anterior = atual
            atual = atual.proximo
            indice += 1

        novo_no.proximo = atual
        anterior.proximo = novo_no

    def remover(self, dado):
        if not self.cabeca:
            return

        if self.cabeca.dado == dado:
            self.cabeca = self.cabeca.proximo
            return

        atual = self.cabeca
        anterior = None

        while atual and atual.dado != dado:
            anterior = atual
            atual = atual.proximo

        if not atual:
            return

        anterior.proximo = atual.proximo

# Criação da lista encadeada e testes dos métodos
lista = ListaEncadeada()
lista.adicionar(1)
lista.adicionar(10)
lista.adicionar(3) 
lista.inserir_no_inicio(2)
lista.remover(10) 

print(lista)